emission/tests/pipelineTests/TestExpectationPipeline.py

```diff
@@ -0,0 +1,115 @@
+import unittest
+import numpy as np
+import time
+import arrow
+
+import emission.storage.timeseries.timequery as estt
+import emission.core.wrapper.labelprediction as ecwl
+import emission.analysis.userinput.expectations as eaue
+import emission.storage.decorations.analysis_timeseries_queries as esda
+import emission.analysis.classification.inference.labels.pipeline as eacilp
+import emission.core.get_database as edb
+import emission.tests.common as etc
+import emission.analysis.configs.expectation_notification_config as eace
+
+class TestExpectationPipeline(unittest.TestCase):
+    test_algorithms = {
+            ecwl.AlgorithmTypes.PLACEHOLDER_3: eacilp.placeholder_predictor_3
+    }
+    tz = "America/Chicago"
+    contrived_dates = {  # Reused from TestExpectationNotificationConfig
+        494: arrow.get("2021-05-01T20:00:00.000", tzinfo=tz),
+        565: arrow.get("2021-06-01T20:00:00.000", tzinfo=tz),
+        795: arrow.get("2021-06-08T20:00:00.000", tzinfo=tz),
+        805: arrow.get("2021-09-04T20:00:00.000", tzinfo=tz),
+        880: arrow.get("2023-02-03T20:00:00.000", tzinfo=tz),
+        960: arrow.get("2023-02-12T20:00:00.000", tzinfo=tz) 
+    }
+
+    def setUp(self):
+        np.random.seed(61297777)
+        self.reset_all()
+        etc.setupRealExample(self, "emission/tests/data/real_examples/shankari_2015-07-22")
+        self.run_pipeline(self.test_algorithms)
+        time_range = estt.TimeQuery("metadata.write_ts", None, time.time())
+        self.inferred_trips = esda.get_entries(esda.INFERRED_TRIP_KEY, self.testUUID, time_query=time_range)
+        self.inferred_id_to_trip = {trip.get_id(): trip for trip in self.inferred_trips}
+        self.expected_trips = esda.get_entries(esda.EXPECTED_TRIP_KEY, self.testUUID, time_query=time_range)
+
+    def tearDown(self):
+        self.reset_all()
+
+    def run_pipeline(self, algorithms):
+        primary_algorithms_stash = eacilp.primary_algorithms
+        eacilp.primary_algorithms = algorithms
+        test_options_stash = eaue._test_options
+        eaue._test_options["preprocess_trip"] = lambda trip: self.preprocess(trip)
+        etc.runIntakePipeline(self.testUUID)
+        eacilp.primary_algorithms = primary_algorithms_stash
+        eaue._test_options = test_options_stash
+    
+    def preprocess(self, trip):
+        # See eacilp.placeholder_predictor_2 for an explanation of the "fingerprint" technique
```

Comment:
I don't see this explanation here - only a one-liner about `Timestamp2index`.

Reply:
I can flesh out that explanation in an upcoming commit.